File: perl/mortgage.pl
#!/usr/bin/perl
use strict;
use Getopt::Std;
use Scalar::Util qw(looks_like_number);
my ($r_m, $d_m, $top_line, $bottom_line, $mp, $ap);
our ($opt_a, $opt_r, $opt_d);
getopt('a:r:d:');
die "Please use -a to input the mortgage amount\n" unless (defined $opt_a);
die "Please use -r to input the annual interest rate\n" unless (defined $opt_r);
die "Please use -d to input the duration of the mortgage in years\n" unless (defined $opt_d);
die "Please ensure that the amount is numeric\n" unless looks_like_number($opt_a);
die "Please ensure that the rate is numeric\n" unless looks_like_number($opt_r);
die "Please ensure that the duration is numeric\n" unless looks_like_number($opt_d);
if ( ( $opt_a == 0 ) || ( $opt_r == 0 ) || ( $opt_d == 0 ) ) 
{
   die "Please ensure that there are no zero values\n";
}

$r_m = $opt_r / 12; ## Interest rate expressed monthky
$d_m = $opt_d * 12; ## Duration in months
$top_line = ( $opt_a * $r_m * ( 1 + ( $r_m / 100 ))**$d_m );
$bottom_line = 100 * (( (1 + ( $r_m / 100 ))**$d_m) -1);
$mp = $top_line / $bottom_line ;
$ap = $mp * 12;

printf("Annual replayments will be £%.2f\n", $ap);
printf("Monthly replayments will be £%.2f\n",$mp);
printf("Total cost of loan will be £%.2f\n", $ap * $opt_d );

File: c/mortgage/mortgage.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
void helptext(char* p_name)
{
   printf("Usage: %s -a <amount> -i <interest rate> -d <duration>\n", p_name);
}
int main(int argc, char** argv)
{
   int amount=0;
   int duration=0;
   double interest_rate, rm, dm;
   double top_line, bottom_line, mp, ap;
   if ( argc < 6 ) {
      helptext(argv[0]);
      exit(-1);
   }
   int i=1;
   while (i < argc ) {
      if ( strcmp(argv[i], "-a")==0 ) {
         i++;
         amount=atoi(argv[i]);
      }
      if ( strcmp(argv[i], "-i")==0 ) {
         i++;
         interest_rate=atof(argv[i]);
      }
      if ( strcmp(argv[i], "-d")==0 ) {
         i++;
         duration=atoi(argv[i]);
      }
      i++;
   }
   if ( amount == 0 || interest_rate == 0 || duration == 0 ) {
      helptext(argv[0]);
      exit(-1);
   }
   printf("Amount=£%d\nInterest=%5.2f%%\nDuration=%d years\n", amount, interest_rate, duration);

   rm = interest_rate / 12;
   dm = duration * 12;

   top_line = ( amount * rm * pow(( 1 + ( rm / 100 )),dm) );
   bottom_line = 100 * ( pow((1 + ( rm / 100 )),dm) -1);

   mp = top_line / bottom_line;
   ap = mp * 12;
   printf("Annual replayments will be £%8.2f\n", ap);
   printf("Monthly replayments will be £%7.2f\n", mp);
   printf("Total cost of loan is £%8.2f\n", (ap * duration));
   exit(0);
}
